File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
def dfs(x, y):
    global cnt
    cnt += 1
    visited[x][y] = True

    for i in range(4):
        nx = x + dx[i]
        ny = y + dy[i]

        if 0 <= nx < N and 0 <= ny < N and arr[nx][ny] == 1 and visited[nx][ny] == False:
            dfs(nx, ny)

N = int(input())
arr = [list(map(int, input())) for _ in range(N)]

visited = [[False] * N for _ in range(N)]

dx = [1, -1, 0, 0]
dy = [0, 0, -1, 1]

cnt = 0
answer = []

for i in range(N):
    for j in range(N):
        if arr[i][j] == 1 and visited[i][j] == False:
            cnt = 0
            dfs(i, j)
            answer.append(cnt)

answer.sort()

print(len(answer))
for i in answer:
    print(i)

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.cc
#include <iostream>
#include <algorithm>
using namespace std;

int N, cnt, apart_cnt, nx, ny;
char input[26];
int apart[626];
int matrix[26][26];
int dx[4] = { 1, 0, -1, 0 };
int dy[4] = { 0, 1, 0, -1 };

int dfs(int x, int y) {
	cnt += 1;

	for (int move = 0; move < 4; move++)
	{
		nx = x + dx[move];
		ny = y + dy[move];

		if ((nx >= 0 || nx < N) && (ny >= 0 || ny < N) && (matrix[nx][ny] == 1)) {
			matrix[nx][ny] = 0;
			dfs(nx, ny);
		}
	}

	return 0;
}

int main(void) {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	
	cin >> N;

	for (int i = 0; i < N; i++)
	{
		cin >> input;

		for (int j = 0; j < N; j++)
		{
			matrix[i][j] = int(input[j]) - int('0');
		}
	}

	for (int i = 0; i < N; i++)
	{
		for (int j = 0; j < N; j++)
		{
			if (matrix[i][j] == 1) {
				cnt = 0;
				matrix[i][j] = 0;
				dfs(i, j);
				apart[apart_cnt] = cnt;
				apart_cnt += 1;
			}
		}
	}

	sort(apart, apart + apart_cnt);

	cout << apart_cnt << "\n";

	for (int i = 0; i < apart_cnt; i++)
	{
		cout << apart[i] << "\n";
	}
}